File: 2016/day23.py
import math
with open("day23.input") as inputfile:
	instructions = [line[:-1].split(' ') for line in inputfile]



#print instructions
pos = 0

#part 1:
registers = { 'a': 8, 'b': 0, 'c': 0, 'd': 0 }

#part 2:
#a! + 5840
#registers = { 'a': 123, 'b': 0, 'c': 1, 'd': 0 }
def toggle(target):
	#print 'target', target
	if target >=0 and target < len(instructions):
		#print 'toggle instruction', instructions[target], 'pos argument:', target
		if len(instructions[target]) == 2:
			if instructions[target][0] == 'inc':
				instructions[target][0] = 'dec'
			else:
				instructions[target][0] = 'inc'
		else:
			if instructions[target][0] == 'jnz':
				instructions[target][0] = 'cpy'
			else:
				instructions[target][0] = 'jnz'


while pos < len(instructions):
	#print instructions[pos]
	#print 'register value', registers['a'], registers['b'], registers['c'], registers['d']
	instruction = instructions[pos][0]
	args = instructions[pos][1:]	
	#print 'instruction', instruction, 'arguments', args
	if instruction == ('cpy'):
		if args[0] not in 'abcd':
			registers[args[1]] = int(args[0])
		else:
			registers[args[1]] = registers[args[0]]
		pos += 1
	elif instruction == 'inc':
		registers[args[0]] += 1
		pos += 1
	elif instruction == 'dec':
		registers[args[0]] -= 1
		pos += 1
	elif instruction == 'jnz':
		if (args[0] not in 'abcd' and args[0] != '0') or registers[args[0]] != 0:
			if args[1] not in 'abcd':
				pos += int(args[1])
			else:
				pos += registers[args[1]]
		else: 
			pos += 1
	elif instruction == 'tgl':
		offset = registers[args[0]]
		#print 'offset:', offset, 'pos', pos
		if offset + pos >= 0 and offset + pos <= len(instructions):
			toggle(pos + offset)
		pos += 1
	elif instruction == 'noop':
		pos += 1
	elif instruction == 'fact':
		f = math.factorial(registers['d'])
		registers['a'] = int(f)
		registers['b'] = 2
		registers['c'] = 0
		registers['d'] = 1
		pos += 1
print registers['a']

File: 2017/day23.go
package main
import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"strconv"
	"math"
)

var answer = 0
func main() {
	solveA("day23.input")
	solveB("day23.input.optimized")
}

/*
func solveC() {
	comps := 0
	for b := 107900; b <= 124900; b += 17 {//loop 1001 {
		if !isPrime(b) {
			comps++
		}
	}
	fmt.Println("composite numbers: ", comps)

}
*/

func getInstructions(inputfile string) map[int]string {
	file, _ := os.Open(inputfile)
        defer file.Close()
        scanner := bufio.NewScanner(file)
	instructions := make(map[int]string)
	i := 0
	for scanner.Scan() {
		instructions[i] = scanner.Text()
		i++;
	}
	return instructions
}


func solveB(file string) {
	instructions := getInstructions(file)
	pos := 0
	registers := make(map[string]int64)
	registers["a"] = 1
	for pos >= 0 {
		pos = doInstruction(instructions[pos], pos, registers)
	}
	fmt.Printf("b: %v\n", registers["h"])
}

func solveA(file string) {
	instructions := getInstructions(file)
	pos := 0
	registers := make(map[string]int64)
	for pos >= 0 {
		pos = doInstruction(instructions[pos], pos, registers)
	}
	fmt.Printf("a: %v\n", answer)
}

func doInstruction(instruction string, pos int, registers map[string] int64) int {
	s := strings.Split(instruction, " ")
	switch(s[0]) {
		case "set": return doSet(s[1], s[2], registers, pos)
		case "sub": return doSub(s[1], s[2], registers, pos)
		case "mul": return doMul(s[1], s[2], registers, pos)
		case "jnz": return doJnz(s[1], s[2], registers, pos)
		case "nop": return pos + 1
		case "prm": return doPrm(s[1], s[2], registers, pos)
		default: {
			fmt.Printf("unknown instruction: %v\n", s);
			return -1
			//panic("crash and burn")
		}
	}
}

func doPrm(register string, value string, registers map[string]int64, pos int) int {
	v := getValue(value, registers)

	if isPrime(int(v)) {
		registers[register] = 1
	} else {
		registers[register] = 0
	}
	return pos + 1
}


func isPrime(value int) bool {
	if value % 2 == 0 {
		return false
	}
	for i := 3; i <= int(math.Floor(math.Sqrt(float64(value)))); i+=2 {
		if value % i == 0 {
			return false
		}
	}
	return value > 1
}


func getValue(value string, registers map[string]int64) int64 {
	v, err := strconv.ParseInt(value, 10, 64)
	if err == nil {
		return v
	} else {
		return registers[value]
	}
}


func doSet(register string, value string, registers map[string]int64, pos int) int {
	v := getValue(value, registers)
	registers[register] = v
	return pos + 1

}

func doSub(register string, value string, registers map[string]int64, pos int) int {
	v := getValue(value, registers)
	registers[register] -= v
	return pos + 1

}

func doMul(register string, value string, registers map[string]int64, pos int) int {
	v := getValue(value, registers)
	registers[register] *= v
	answer++
	return pos + 1
}

func doJnz(register string, value string, registers map[string]int64, pos int) int {
	r := getValue(register, registers)
	v := getValue(value, registers)
	if r != 0 {
		return pos + int(v)
	} else {
		return pos + 1
	}
}
